lib/linguist/heuristics.yml

```diff
@@ -268,6 +268,13 @@ disambiguations:
     named_pattern: perl6
   - language: XPM
     pattern: '^\s*\/\* XPM \*\/'
+- extensions: ['.pp']
+  rules:
+  - language: Pascal
+    pattern: '^\s*end[.;]'
+  - language: Puppet
+    # Curly brackets mark comments in Pascal.
```

Comment:
Pascal comments can span multiple lines, according to [Free Pascal's wiki](http://wiki.freepascal.org/Comments) (going by the distinction the page makes between block and line comments).

So a Pascal file with a multiline header might get flagged as Puppet.

Reply:
I was [under the impression](https://www.tutorialspoint.com/pascal/pascal_basic_syntax.htm) that multiline comments had to start and end with `{*` and `*}`. Is that incorrect?

Reply:
Looks like I was wrong: https://github.com/github/linguist/blob/b70a976e12d52184d91780db2607d8152f524e88/samples/Pascal/cwindirs.pp#L77-L79

Reply:
It might be safer to assert `Puppet` by matching a keyword preceding the opening brace. To use the [examples here](https://puppet.com/docs/puppet/5.3/lang_summary.html), the heuristic might look like:

~~~regexp
/^[ \t]+(package|service|file)[ \t]+}$/
~~~

Reply:
BTW, it appears the author of [that page you linked](https://www.tutorialspoint.com/pascal/pascal_basic_syntax.htm) is misinformed about comment syntax — the [Free Pascal Wiki](http://wiki.freepascal.org/Comments) explains that `(* … *)` is a relic from the old days when keyboards didn't always have curly brackets (!). The single-line comment form is `// `, just like C++. =)

So the asterisks in that page's example are redundant...

Reply:
Ah yes, sorry. 😅 That's what I meant. Fixed.

Reply:
> BTW, it appears the author of [that page you linked](https://www.tutorialspoint.com/pascal/pascal_basic_syntax.htm) is misinformed about comment syntax — the [Free Pascal Wiki](http://wiki.freepascal.org/Comments) explains that `(* … *)` is a relic from the old days when keyboards didn't always have curly brackets (!). The single-line comment form is `// `, just like C++. =)

To be fair, there are many Pascal dialects and even more implementations, so both statements could be true. We need to support both cases though, so I'll change the heuristic rule.

Reply:
That heuristic matches only 1 of our 5 sample files. They're not very big, but still.
Maybe
```regexp
/^\s+\w+\s+=>\s/
```
would be a better match?

Reply:
Assuming `=>` isn't legal syntax in most Pascal dialects (you seem to know more here than I do), I'd say that's an acceptable heuristic. :+1:

Reply:
I don't know much on Pascal either; only what I remember from classes a decade ago...

To check, I downloaded 4,957 `.pascal` and `.pas` Pascal files and searched for the above regular expression. None of the files contain it so I'm fairly confident that, if  `=>` is legal syntax, it's very uncommon.